File: java/main/com/facebook/profilo/logger/MultiBufferLogger.java
/**
 * Copyright 2004-present, Facebook, Inc.
 *
 * <p>Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of the License at
 *
 * <p>http://www.apache.org/licenses/LICENSE-2.0
 *
 * <p>Unless required by applicable law or agreed to in writing, software distributed under the
 * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.facebook.profilo.logger;

import com.facebook.jni.HybridClassBase;
import com.facebook.profilo.core.ProfiloConstants;
import com.facebook.profilo.mmapbuf.core.Buffer;
import com.facebook.proguard.annotations.DoNotStripAny;
import com.facebook.soloader.SoLoader;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * MultiBufferLogger allows entries to be written to one or more buffers at the same time.
 *
 * <p>It's important to note that while the Java-side API relies on BufferLogger for the actual
 * writes, there's a complementary C++ API as well.
 */
@DoNotStripAny
public final class MultiBufferLogger extends HybridClassBase {

  private volatile boolean mLoaded;
  private final AtomicInteger mBuffersCount;
  private volatile long mNativePtr;

  public MultiBufferLogger() {
    mBuffersCount = new AtomicInteger(0);
  }

  private void ensureLoaded() {
    boolean loaded = mLoaded;
    if (loaded) {
      return;
    }

    synchronized (this) {
      loaded = mLoaded;
      if (loaded) {
        return;
      }

      SoLoader.loadLibrary("profilo");
      mNativePtr = initHybrid();
      mLoaded = true;
    }
  }

  public void addBuffer(Buffer buffer) {
    ensureLoaded();
    nativeAddBuffer(buffer);
    mBuffersCount.incrementAndGet();
  }

  public void removeBuffer(Buffer buffer) {
    ensureLoaded();
    nativeRemoveBuffer(buffer);
    mBuffersCount.decrementAndGet();
  }

  public int writeStandardEntry(
      int flags,
      int type,
      long timestamp,
      int tid,
      int arg1 /* callid */,
      int arg2 /* matchid */,
      long arg3 /* extra */) {
    if (mBuffersCount.intValue() == 0) {
      return ProfiloConstants.NO_MATCH;
    }
    ensureLoaded();
    return nativeWriteStandardEntry(mNativePtr, flags, type, timestamp, tid, arg1, arg2, arg3);
  }

  public int writeBytesEntry(int flags, int type, int arg1 /* matchid */, String arg2 /* bytes */) {
    if (mBuffersCount.intValue() == 0) {
      return ProfiloConstants.NO_MATCH;
    }
    ensureLoaded();
    return nativeWriteBytesEntry(mNativePtr, flags, type, arg1, arg2);
  }

  private static native int nativeWriteStandardEntry(
      long ptr,
      int flags,
      int type,
      long timestamp,
      int tid,
      int arg1 /* callid */,
      int arg2 /* matchid */,
      long arg3 /* extra */);

  private static native int nativeWriteBytesEntry(
      long ptr, int flags, int type, int arg1 /* matchid */, String arg2 /* bytes */);

  private native long initHybrid();

  private native void nativeAddBuffer(Buffer buffer);

  private native void nativeRemoveBuffer(Buffer buffer);
}

File: cpp/logger/MultiBufferLogger.cpp
/**
 * Copyright 2004-present, Facebook, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "MultiBufferLogger.h"

namespace facebook {
namespace profilo {
namespace logger {

MultiBufferLogger::MultiBufferLogger(MultiBufferLogger::EntryIDCounter& counter)
    : entryID_(counter) {}

void MultiBufferLogger::addBuffer(std::shared_ptr<Buffer> buffer) {
  WriterLock lock(&lock_);
  buffers_.push_back(buffer);
}

void MultiBufferLogger::removeBuffer(std::shared_ptr<Buffer> buffer) {
  WriterLock lock(&lock_);
  auto iter = std::find(buffers_.begin(), buffers_.end(), buffer);
  if (iter == buffers_.end()) {
    return;
  }
  buffers_.erase(iter);
}

int32_t MultiBufferLogger::writeBytes(
    EntryType type,
    int32_t arg1,
    const uint8_t* arg2,
    size_t len) {
  if (len > Logger::kMaxVariableLengthEntry) {
    throw std::overflow_error("len is bigger than kMaxVariableLengthEntry");
  }
  if (arg2 == nullptr) {
    throw std::invalid_argument("arg2 is null");
  }

  // Maintain the same entry ID across all buffers
  auto id = entryID_.next();

  BytesEntry entry{
      .id = id,
      .type = type,
      .matchid = arg1,
      .bytes =
          {
              .values = arg2,
              .size = static_cast<uint16_t>(len),
          },
  };

  ReaderLock lock(&lock_);
  for (auto& buf : buffers_) {
    buf->logger().write(entry);
  }
  return entry.id;
}
} // namespace logger
} // namespace profilo
} // namespace facebook
